### PR TITLE
pm: power: npcx: add console expired mechanism.

### PR DESCRIPTION
**Pull request by MulinChao**
This CL adds support for console expired mechanism. It implements the
notification to power management module that the module for console is
in use. If the interval that module doesn't receive any input message
exceeds CONFIG_SOC_POWER_CONSOLE_EXPIRED_TIMEOUT, the power management
module is allowed to enter deep sleep mode. This mechanism gives a
window in which the users can organize console input.

By npcx_power_console_is_in_use(), the application can forbid ec enter deep sleep in customized pm_policy_next_state().

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

### CODE REVIEW

**Comment by MulinChao** [comment]
Sorry for the noise. This is a draft version. Still need more time to verify its stability.

**Comment by MulinChao** [comment]
Passed the verification on volteer with [PR32873](https://github.com/zephyrproject-rtos/zephyr/pull/32873) ,[CL2721322](https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/2721322), and [CL2738500](https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/2738500).   
Ready for code review now.

File: soc/arm/nuvoton_npcx/common/soc_power.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+bool npcx_power_console_is_in_use(void);
+
+/**
+ * @brief Notify the power module that the module for the console is in use.
+ *
+ * Notify the power module that the module for the console is in use. It also
+ * extends expired time by CONFIG_SOC_POWER_CONSOLE_EXPIRED_TIMEOUT.
+ */
+void npcx_power_console_is_in_use_refresh(void);
+
```
Comment (keith-zephyr) [inline soc/arm/nuvoton_npcx/common/soc_power.h:31]: Can you also add an API for changing the console activity timeout?   See https://chromium.googlesource.com/chromiumos/platform/ec/+/cfbbbb93/chip/npcx/clock.c#474 for the expected usage.

File: soc/arm/nuvoton_npcx/npcx7/Kconfig.soc [hunk 2]
```diff
@@ -33,3 +33,23 @@ config SOC_POWER_MANAGEMENT_TRACE
...
+	  is allowed to enter deep sleep mode. This mechanism gives a window in
+	  which the users can organize console input.
+
+config SOC_POWER_CONSOLE_EXPIRED_TIMEOUT
+	int "Fixed amount of time to keep the console in use flag true"
+	default 15000
+	depends on SOC_POWER_CONSOLE_EXPIRED
+	help
+	  Fixed amount of time which unit is milli-second to keep the console in
+	  use flag true.
```
Comment (keith-zephyr) [inline soc/arm/nuvoton_npcx/npcx7/Kconfig.soc:55]: Can these configs get promoted to work with the general CONFIG_UART_CONSOLE?

Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/Kconfig.soc]: Do you mean that?
1. Replace the prefix "SOC_POWER_CONSOLE_" with "UART_CONSOLE_".
2. Move these two configs to [console driver's Kconfig file](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/console/Kconfig#L47)

If so, I will modify it later.

File: soc/arm/nuvoton_npcx/common/soc_power.h (hunk 1 above)
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/soc_power.h:31]: Done.

File: soc/arm/nuvoton_npcx/npcx7/Kconfig.soc (hunk 2 above)
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/Kconfig.soc]: Done.

File: drivers/serial/uart_npcx.c [hunk 3]
```diff
@@ -209,6 +212,12 @@ static void uart_npcx_isr(const struct device *dev)
 {
 	struct uart_npcx_data *data = DRV_DATA(dev);
 
+	/* Refresh console expired time if got UART Rx event */
+	if (IS_ENABLED(CONFIG_UART_CONSOLE_INPUT_EXPIRED) &&
+						uart_npcx_irq_rx_ready(dev)) {
```
Comment (jackrosenthal) [inline drivers/serial/uart_npcx.c:217]: line seems over-indented (should align to left-paren of if)

File: drivers/serial/uart_npcx.c [hunk 4]
```diff
@@ -367,6 +386,10 @@ static int uart_npcx_init(const struct device *dev)
 #endif
 
 #if defined(CONFIG_PM)
+	/* Initialize a miwu device input and its callback function */
+	npcx_miwu_init_dev_callback(&data->uart_rx_cb, &config->uart_rx_wui,
+						uart_npcx_rx_wk_isr, dev);
```
Comment (jackrosenthal) [inline drivers/serial/uart_npcx.c:391]: line is over-indented (should align to left-paren of the function call)

File: drivers/serial/uart_npcx.c (hunk 3 above)
Comment (MulinChao) [reply drivers/serial/uart_npcx.c]: Done.

File: drivers/serial/uart_npcx.c (hunk 4 above)
Comment (MulinChao) [reply drivers/serial/uart_npcx.c]: Done.

**Review by jackrosenthal — APPROVED** [review]

File: drivers/console/Kconfig [hunk 5]
```diff
@@ -81,6 +81,27 @@ config UART_CONSOLE_MCUMGR
...
+	  allowed to enter sleep/deep sleep state and turn off the clock of UART
+	  console module. This mechanism gives a window in which the users can
+	  organize input message if CONFIG_PM is enabled.
+
+config UART_CONSOLE_INPUT_EXPIRED_TIMEOUT
+	int "Fixed amount of time to keep the UART console in use flag true"
+	default 15000
+	depends on UART_CONSOLE_INPUT_EXPIRED
+	help
+	  Fixed amount of time which unit is milli-second to keep the UART
```
Comment (sjg20) [inline drivers/console/Kconfig:102]: milliseconds
or ms

**Review by sjg20 — APPROVED** [review]

File: drivers/serial/uart_npcx.c [hunk 6]
```diff
@@ -38,6 +38,9 @@ struct uart_npcx_data {
 	uart_irq_callback_user_data_t user_cb;
 	void *user_data;
 #endif
+#ifdef CONFIG_PM
```
Comment (sjg20) [inline drivers/serial/uart_npcx.c:41]: Can this #ifdef be dropped?

File: drivers/serial/uart_npcx.c [hunk 7]
```diff
@@ -367,6 +386,10 @@ static int uart_npcx_init(const struct device *dev)
 #endif
 
 #if defined(CONFIG_PM)
```
Comment (sjg20) [inline drivers/serial/uart_npcx.c:388]: Really would like to use IS_ENABLED() if possible?

**Review by sjg20 — CHANGES_REQUESTED** [review]

File: drivers/console/Kconfig (hunk 5 above)
Comment (MulinChao) [reply drivers/console/Kconfig]: Done.

File: drivers/serial/uart_npcx.c (hunk 6 above)
Comment (MulinChao) [reply drivers/serial/uart_npcx.c:41]: Please see the reply in line 388 below.

File: drivers/serial/uart_npcx.c (hunk 7 above)
Comment (MulinChao) [reply drivers/serial/uart_npcx.c:388]: In this section, the driver registers a callback function, [uart_npcx_rx_wk_isr](https://github.com/zephyrproject-rtos/zephyr/pull/32776/files#diff-80314ccee38f06d27252b7ac2d69a582d3687c8b71808737fdbbb0bc31182fd9R298), in [line 391](https://github.com/zephyrproject-rtos/zephyr/pull/32776/files#diff-80314ccee38f06d27252b7ac2d69a582d3687c8b71808737fdbbb0bc31182fd9R391) for handling the related wake-up input event. If this PR replaces "#ifdef" with "IS_ENABLED()" here, we will get a linking error, "'uart_npcx_rx_wk_isr' undeclared", if CONFIG_PM is not enabled. That's why I didn't choose IS_ENABLED() here.

File: drivers/serial/uart_npcx.c [hunk 8]
```diff
@@ -285,6 +294,16 @@ static int uart_npcx_err_check(const struct device *dev)
 	return err;
 }
 
+#ifdef CONFIG_PM
+static void uart_npcx_rx_wk_isr(const struct device *dev, struct npcx_wui *wui)
```
Comment (keith-zephyr) [inline drivers/serial/uart_npcx.c:298]: Replace #ifdef CONFIG_PM with __unused static void uart_npcx_rx_wk_isr().   Then you can also use IS_ENABLED() at line 369.

Comment (MulinChao) [reply drivers/serial/uart_npcx.c:298]: Oh, yes. Thanks for reminding me. I will submit the PR with this change later.

File: drivers/serial/uart_npcx.c (hunk 6 above)
Comment (MulinChao) [reply drivers/serial/uart_npcx.c]: Done.

File: drivers/serial/uart_npcx.c (hunk 7 above)
Comment (MulinChao) [reply drivers/serial/uart_npcx.c]: Done.

File: drivers/serial/uart_npcx.c (hunk 8 above)
Comment (MulinChao) [reply drivers/serial/uart_npcx.c]: Done.

**Review by keith-zephyr — APPROVED** [review]

**Review by sjg20 — APPROVED** [review]

**Pull request merged**